src/olmo_core/train/callbacks/gradient_dumper.py

```diff
@@ -0,0 +1,111 @@
+        step_dir.mkdir(exist_ok=True, parents=True)
+
+        assert hasattr(self.trainer.train_module, "model")
+        model = getattr(self.trainer.train_module, "model")
+
+        if self.save_first_n is None:
+            # Save full gradients using distributed checkpoint
+            full_grads_dir = step_dir / "full_gradients"
+
+            grad_dict = {}
+            for name, p in model.named_parameters():
+                if p.grad is not None:
+                    grad_dict[name] = p.grad.detach()
+
+            log.info(f"Saving {len(grad_dict)} gradient tensors for step {self.step}...")
+            save_state_dict(
+                full_grads_dir,
+                grad_dict,
+                save_overwrite=True,
+            )
+            log.info(f"Saved full gradients for step {self.step} to '{full_grads_dir}'")
+        else:
+            # Save first N elements of full gradients, gather all gradients and save from rank 0
+            sampled_gradients_dir = step_dir / "sampled_gradients"
+            # Only rank 0 creates the directory and logs summary since only rank 0 writes files
+            if get_rank() == 0:
+                sampled_gradients_dir.mkdir(exist_ok=True, parents=True)
+
+            for name, p in model.named_parameters():
+                if p.grad is None:
+                    continue
+
+                # get_full_tensor handles DTensor all-gather automatically
+                full_grad = get_full_tensor(p.grad.detach())
+
+                # Only rank 0 saves
+                if get_rank() == 0:
+                    full_grad = full_grad.cpu()
+
+                    # Slice first N elements along dimension 0
+                    dim_size = full_grad.shape[0]
+                    actual_n = min(self.save_first_n, dim_size)
+                    if actual_n < self.save_first_n:
+                        log.warning(
+                            f"Parameter '{name}': save_first_n={self.save_first_n} exceeds "
+                            f"dimension size {dim_size}, capping to {actual_n}"
+                        )
+
+                    sliced_grad = full_grad.narrow(0, 0, actual_n)
+                    slice_filename = f"{name}_first{actual_n}.safetensors"
+                    slice_filepath = sampled_gradients_dir / slice_filename
+                    save_file({"gradient": sliced_grad}, str(slice_filepath))
+                    log.info(f"Saved first {actual_n} of '{name}' to '{slice_filepath}'")
+
+                del full_grad
+
+            if get_rank() == 0:
+                log.info(
+                    f"Saved sampled gradients for step {self.step} to '{sampled_gradients_dir}'"
+                )
+
+        if get_rank() == 0:
+            rel_step_dir = step_dir.relative_to(self.trainer.work_dir)
+            target_dir = self.trainer.persist_working_subdir(rel_step_dir)
+            log.info(f"Gradients for step {self.step} uploaded to '{target_dir}'")
```

Comment:
The save folder could be local.
```suggestion
            log.info(f"Gradients for step {self.step} saved to '{target_dir}'")
```